File: practice-projects/gin-ddd/internal/domain/event/domain_event.go
// Package event 定义领域事件与发布/订阅端口。
//
// 领域事件表达"已发生的事实",字段公开符合 Go 习惯,但仅由各事件工厂函数
// 构造,创建后不应再修改——这是领域事件不可变性的契约约定。
package event

import "time"

// DomainEvent 领域事件接口。
type DomainEvent interface {
	// EventType 事件类型(过去时态语义)。
	EventType() string

	// OccurredOn 事件发生时间。
	OccurredOn() time.Time

	// EventData 事件数据载荷。
	EventData() interface{}
}

// BaseEvent 基础事件,提供事件类型与时间戳的公共字段。
//
// 字段公开仅供 JSON 序列化使用,业务代码不应直接构造或修改 BaseEvent,
// 而应通过具体事件类型的 NewXxxEvent 工厂函数。
type BaseEvent struct {
	Type      string    `json:"type"`
	Timestamp time.Time `json:"timestamp"`
}

// EventType 返回事件类型。
func (e *BaseEvent) EventType() string { return e.Type }

// OccurredOn 返回事件发生时间。
func (e *BaseEvent) OccurredOn() time.Time { return e.Timestamp }

File: domain-driven-design/java-web/src/main/java/com/microwind/javaweborder/domain/event/DomainEvent.java
package com.microwind.javaweborder.domain.event;

import java.time.LocalDateTime;

/**
 * 领域事件通用契约。
 *
 * <p>DDD 战术构件：<b>领域事件（Domain Event）</b>。当业务上发生某件
 * "值得记录的事"（订单创建、订单取消等）时，由聚合根累积事件，
 * 由应用层在事务边界外通过 {@link DomainEventPublisher} 一次性发布，
 * 用于解耦不同聚合、不同限界上下文之间的协作。
 *
 * <p>命名约定：具体事件类以"过去时"命名（OrderCreated、OrderCanceled），
 * 表示"已经发生的事实"，不可被否决或回滚。
 */
public interface DomainEvent {

    /**
     * @return 事件发生时间（构造时固化）
     */
    LocalDateTime occurredOn();

    /**
     * @return 事件类型标识，用于发布 / 订阅 / 日志聚合
     */
    String eventType();
}
